homeassistant/components/homekit/type_air_purifiers.py

```diff
@@ -0,0 +1,469 @@
+"""Class to hold all air purifier accessories."""
+import logging
+
+from pyhap.const import CATEGORY_AIR_PURIFIER
+
+from homeassistant.components.fan import (
+    ATTR_DIRECTION,
+    ATTR_OSCILLATING,
+    ATTR_PERCENTAGE,
+    ATTR_PERCENTAGE_STEP,
+    ATTR_PRESET_MODE,
+    ATTR_PRESET_MODES,
+    DIRECTION_FORWARD,
+    DIRECTION_REVERSE,
+    DOMAIN,
+    SERVICE_OSCILLATE,
+    SERVICE_SET_DIRECTION,
+    SERVICE_SET_PERCENTAGE,
+    SERVICE_SET_PRESET_MODE,
+    FanEntityFeature,
+)
+from homeassistant.const import (
+    ATTR_ENTITY_ID,
+    ATTR_SUPPORTED_FEATURES,
+    SERVICE_TURN_OFF,
+    SERVICE_TURN_ON,
+    STATE_OFF,
+    STATE_ON,
+)
+from homeassistant.core import callback
+from homeassistant.helpers.event import async_track_state_change_event
+
+from .accessories import TYPES, HomeAccessory
+from .const import (
+    CHAR_ACTIVE,
+    CHAR_AIR_QUALITY,
+    CHAR_CURRENT_AIR_PURIFIER_STATE,
+    CHAR_CURRENT_HUMIDITY,
+    CHAR_CURRENT_TEMPERATURE,
+    CHAR_NAME,
+    CHAR_PM25_DENSITY,
+    CHAR_ROTATION_DIRECTION,
+    CHAR_ROTATION_SPEED,
+    CHAR_SWING_MODE,
+    CHAR_TARGET_AIR_PURIFIER_STATE,
+    CONF_LINKED_HUMIDITY_SENSOR,
+    CONF_LINKED_PM25_SENSOR,
+    CONF_LINKED_TEMPERATURE_SENSOR,
+    PROP_MIN_STEP,
+    SERV_AIR_PURIFIER,
+    SERV_AIR_QUALITY_SENSOR,
+    SERV_HUMIDITY_SENSOR,
+    SERV_TEMPERATURE_SENSOR,
+)
+from .util import cleanup_name_for_homekit, density_to_air_quality
+
+_LOGGER = logging.getLogger(__name__)
+
+CURRENT_STATE_INACTIVE = 0
+CURRENT_STATE_IDLE = 1
+CURRENT_STATE_PURIFYING_AIR = 2
+
+
+@TYPES.register("AirPurifier")
+class AirPurifier(HomeAccessory):
```

Comment:
I think this would be easier to maintain if you made a base fan class and this inherited from it so we don't have mostly duplicate code between here and the fan class

Reply:
Thanks! I’ll hopefully have time to look into that at the weekend.

Reply:
Now done.  Please take a look at the revised version and let me know if I should make further changes.